File: go-server/rabbitMQ/rabbitMQ_helpers.py
#!/usr/bin/env python
import os
import pika
import sys

RABBITMQ_HOST = os.environ.get("RABBITMQ_HOST")
if not RABBITMQ_HOST:
    raise EnvironmentError("RABBITMQ_HOST environment variable is not set.")
RABBITMQ_PORT = os.environ.get("RABBITMQ_PORT")
if not RABBITMQ_PORT:
    raise EnvironmentError("RABBITMQ_PORT environment variable is not set.")
RABBITMQ_USER = os.environ.get("RABBITMQ_USER")
if not RABBITMQ_USER:
    raise EnvironmentError("RABBITMQ_USER environment variable is not set.")
RABBITMQ_PASS = os.environ.get("RABBITMQ_PASS")
if not RABBITMQ_PASS:
    raise EnvironmentError("RABBITMQ_PASS environment variable is not set.")
EXCHANGE_NAME = os.environ.get("RABBITMQ_EXCHANGE_NAME")
if not EXCHANGE_NAME:
    raise EnvironmentError("RABBITMQ_EXCHANGE_NAME environment variable is not set.")


def producer_connection_init():
    """
        Initialize the connection to RabbitMQ and return the channel and the connection.
    """
    
    connection = pika.BlockingConnection(
        pika.ConnectionParameters(host=RABBITMQ_HOST, port=int(RABBITMQ_PORT), credentials=pika.PlainCredentials(RABBITMQ_USER, RABBITMQ_PASS)))
    channel = connection.channel()

    channel.exchange_declare(exchange=EXCHANGE_NAME, exchange_type='topic', durable=True)

    return channel, connection

def publish_message(channel, topic: str, body: str):
    channel.basic_publish(
    exchange=EXCHANGE_NAME, routing_key=topic, body=body)

def connection_end(connection, channel):
    """
        Close the connection to RabbitMQ.
    """
    channel.close()
    connection.close()

def listen(topic: str, callback):
    """
        :param str topic: The topic from which to consume
        :param callable on_message_callback: Required function to handle incoming messages, having the signature:
            on_message_callback(channel, method, properties, body)
            - channel: BlockingChannel
            - method: spec.Basic.Deliver
            - properties: spec.BasicProperties
            - body: bytes
    """
    connection = pika.BlockingConnection(
        pika.ConnectionParameters(host=RABBITMQ_HOST, port=int(RABBITMQ_PORT), credentials=pika.PlainCredentials(RABBITMQ_USER, RABBITMQ_PASS)))
    channel = connection.channel()

    channel.exchange_declare(exchange=EXCHANGE_NAME, exchange_type='topic', durable=True)

    result = channel.queue_declare('', exclusive=True)
    queue_name = result.method.queue

    channel.queue_bind(
        exchange=EXCHANGE_NAME, queue=queue_name, routing_key=topic)

    print(' [*] Waiting for logs. To exit press CTRL+C')

    channel.basic_consume(
        queue=queue_name, on_message_callback=callback, auto_ack=True)

    channel.start_consuming()
    

File: go-server/rabbitMQ/rabbitMQ_helpers.go
package rabbitMQ

import (
	"context"
	"fmt"
	"log"
	"os"
	"strconv"
	"time"

	amqp "github.com/rabbitmq/amqp091-go"
)

var (
	rabbitMQHost = mustGetEnv("RABBITMQ_HOST")
	rabbitMQPort = mustGetEnvInt("RABBITMQ_PORT")
	rabbitMQUser = mustGetEnv("RABBITMQ_USER")
	rabbitMQPass = mustGetEnv("RABBITMQ_PASS")
	exchangeName = mustGetEnv("RABBITMQ_EXCHANGE_NAME")
)

func mustGetEnv(key string) string {
	value := os.Getenv(key)
	if value == "" {
		log.Fatalf("Environment variable %s is required but not set", key)
	}
	return value
}

func mustGetEnvInt(key string) int {
	value := os.Getenv(key)
	if value == "" {
		log.Fatalf("Environment variable %s is required but not set", key)
	}
	v, err := strconv.Atoi(value)
	if err != nil {
		log.Fatalf("Environment variable %s must be an integer, got: %s", key, value)
	}
	return v
}

type Producer struct {
	conn    *amqp.Connection
	channel *amqp.Channel
	ctx     context.Context
	cancel  context.CancelFunc
}

func failOnError(err error, msg string) {
	if err != nil {
		log.Panicf("%s: %s", msg, err)
	}
}

func ProducerConnexionInit() *Producer {
	conn, err := amqp.Dial(fmt.Sprintf("amqp://%s:%s@%s:%d/", rabbitMQUser, rabbitMQPass, rabbitMQHost, rabbitMQPort))
	failOnError(err, "Failed to connect to RabbitMQ")

	ch, err := conn.Channel()
	failOnError(err, "Failed to open a channel")

	err = ch.ExchangeDeclare(
		exchangeName, // name
		"topic",      // type
		true,         // durable
		false,        // auto-deleted
		false,        // internal
		false,        // no-wait
		nil,          // arguments
	)
	failOnError(err, "Failed to declare an exchange")

	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	return &Producer{conn, ch, ctx, cancel}
}

func PublishMessage(p *Producer, body, topic string) {
	err := p.channel.PublishWithContext(p.ctx,
		exchangeName, // exchange
		topic,        // routing key
		false,        // mandatory
		false,        // immediate
		amqp.Publishing{
			ContentType: "text/plain",
			Body:        []byte(body),
		})
	failOnError(err, "Failed to publish a message")

	log.Printf(" [%s] Sent %s", topic, body)
}

func (p *Producer) ConnexionEnd() {
	if err := p.conn.Close(); err != nil {
		log.Printf("Failed to close connection: %s", err)
	}
	if err := p.channel.Close(); err != nil {
		log.Printf("Failed to close channel: %s", err)
	}
	p.cancel()
}

func Listen(topic string, consumeAction func(amqp.Delivery)) {
	conn, err := amqp.Dial(fmt.Sprintf("amqp://%s:%s@%s:%d/", rabbitMQUser, rabbitMQPass, rabbitMQHost, rabbitMQPort))
	failOnError(err, "Failed to connect to RabbitMQ")
	defer conn.Close()

	ch, err := conn.Channel()
	failOnError(err, "Failed to open a channel")
	defer ch.Close()

	err = ch.ExchangeDeclare(
		exchangeName, // name
		"topic",      // type
		true,         // durable
		false,        // auto-deleted
		false,        // internal
		false,        // no-wait
		nil,          // arguments
	)
	failOnError(err, "Failed to declare an exchange")

	q, err := ch.QueueDeclare(
		"",    // name
		false, // durable
		false, // delete when unused
		true,  // exclusive
		false, // no-wait
		nil,   // arguments
	)
	failOnError(err, "Failed to declare a queue")

	log.Printf("Binding queue %s to exchange %s with routing key %s",
		q.Name, exchangeName, topic)
	err = ch.QueueBind(
		q.Name,       // queue name
		topic,        // routing key
		exchangeName, // exchange
		false,
		nil)
	failOnError(err, "Failed to bind a queue")

	msgs, err := ch.Consume(
		q.Name, // queue
		"",     // consumer
		true,   // auto ack
		false,  // exclusive
		false,  // no local
		false,  // no wait
		nil,    // args
	)
	failOnError(err, "Failed to register a consumer")

	var forever chan struct{}

	go func() {
		for d := range msgs {
			consumeAction(d)
		}
	}()

	log.Printf(" [*] Waiting for logs. To exit press CTRL+C")
	<-forever
}
